File: allcode/3100-3199/3197minimumSum.py
# 给你一个二维 二进制 数组 grid。你需要找到 3 个 不重叠、面积 非零 、边在水平方向和竖直方向上的矩形，并且满足 grid 中所有的 1 都在这些矩形的内部。
#
# 返回这些矩形面积之和的 最小 可能值。
#
# 注意，这些矩形可以相接。
#
#
#
# 示例 1：
#
# 输入： grid = [[1,0,1],[1,1,1]]
#
# 输出： 5
#
# 解释：
#
#
#
# 位于 (0, 0) 和 (1, 0) 的 1 被一个面积为 2 的矩形覆盖。
# 位于 (0, 2) 和 (1, 2) 的 1 被一个面积为 2 的矩形覆盖。
# 位于 (1, 1) 的 1 被一个面积为 1 的矩形覆盖。
# 示例 2：
#
# 输入： grid = [[1,0,1,0],[0,1,0,1]]
#
# 输出： 5
#
# 解释：
#
#
#
# 位于 (0, 0) 和 (0, 2) 的 1 被一个面积为 3 的矩形覆盖。
# 位于 (1, 1) 的 1 被一个面积为 1 的矩形覆盖。
# 位于 (1, 3) 的 1 被一个面积为 1 的矩形覆盖。
#
#
# 提示：
#
# 1 <= grid.length, grid[i].length <= 30
# grid[i][j] 是 0 或 1。
# 输入保证 grid 中至少有三个 1 。

from leetcode.allcode.competition.mypackage import *

class Solution:
    def minimumSum(self, grid: List[List[int]]) -> int:
        r, c = len(grid), len(grid[0])

        def minimumArea(r1, r2, c1, c2) -> int:
            # 求一个矩形区域内，包含所有1的最小矩形，没有1时，返回1
            left, right, up, down = inf, -inf, -inf, inf
            for i in range(r1, r2):
                for j in range(c1, c2):
                    if grid[i][j]:
                        left = min(left, j)
                        right = max((right, j))
                        up = max(up, i)
                        down = min(down, i)
            if left == inf:
                return 1
            return (right - left + 1) * (up - down + 1)

        ans = inf

        def f():
            nonlocal ans
            # 上中下三个矩形
            for i in range(1, r):
                for j in range(i + 1, r):
                    ans = min(ans, minimumArea(0, i, 0, c) + minimumArea(i, j, 0, c) + minimumArea(j, r, 0, c))
            # 上/左下/右下 三个矩形
            for i in range(1, r):
                for j in range(1, c):
                    ans = min(ans, minimumArea(0, i, 0, c) + minimumArea(i, r, 0, j) + minimumArea(i, r, j, c))
            # 左上/右上/下 三个矩形
            for i in range(1, r):
                for j in range(1, c):
                    ans = min(ans, minimumArea(0, i, 0, j) + minimumArea(0, i, j, c) + minimumArea(i, r, 0, c))

        f()
        grid = list(zip(*grid[::-1]))
        r, c = len(grid), len(grid[0])
        f()
        return ans



so = Solution()
print(so.minimumSum([[1,0,1],[1,1,1]]))
print(so.minimumSum([[1,0,1,0],[0,1,0,1]]))

File: c++/src/3000-3999/3197minimumSum.cpp
#include "lc_pub.h"


class Solution {
    public:
    int minimumSum(vector<vector<int>>& grid) {
        int r=grid.size(),c=grid[0].size();
        auto calc = [&] (int a,int b,int c, int d) -> int {
            if (a>b||c>d) return 0;
            int up=b,down=a,l=d,ri=c;
            for (int i=a;i<=b;i++)
                for (int j=c;j<=d;j++)
                    if (grid[i][j]) {
                        up=min(up,i);down=max(down,i);l=min(l,j);ri=max(ri,j);
                    }
            return max((down-up+1)*(ri-l+1),0);
        };
        int ans=r*c;
        for (int i=0;i<r;i++) {
            int s1=calc(0,i,0,c-1);
            for (int j=i+1;j<r;j++) {
                int s2=calc(i+1,j,0,c-1);
                int s3=calc(j+1,r-1,0,c-1);
                ans=min(ans,s1+s2+s3);
            }
            for (int j=0;j<c;j++) {
                int s2=calc(i+1,r-1,0,j);
                int s3=calc(i+1,r-1,j+1,c-1);
                ans=min(ans,s1+s2+s3);
            }
            s1=calc(i+1,r-1,0,c-1);
            for (int j=0;j<c;j++) {
                ans=min(ans,s1+calc(0,i,0,j)+calc(0,i,j+1,c-1));
            }
        }
        for (int i=0;i<c;i++) {
            int s1=calc(0,r-1,0,i);
            for (int j=i+1;j<c;j++) {
                ans=min(ans,s1+calc(0,r-1,i+1,j)+calc(0,r-1,j+1,c-1));
            }
            for (int j=0;j<r;j++) {
                ans=min(ans,s1+calc(0,j,i+1,c-1)+calc(j+1,r-1,i+1,c-1));
            }
            s1=calc(0,r-1,i+1,c-1);
            for (int j=0;j<r;j++) {
                ans=min(ans,s1+calc(0,j,0,i)+calc(j+1,r-1,0,i));
            }
        }
        return ans;
    }
    };

    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    auto nums=parseGrid("[[0,0,0],[0,0,0],[0,0,1],[1,1,0]]");

    Solution so;
    cout<<so.minimumSum(nums)<<endl;
    return 0;
}
